File: Yandex/Practicum/Algorightms_and_Data_Structures/8_sprint/G._Поиск_со_сдвигом/solution.py
import sys


def solution() -> None:
    """
    Time Complexity O(NM)
    Memory Complexity O(1)
    """
    n: int = int(sys.stdin.readline().rstrip())
    a: list[int] = list(map(int, sys.stdin.readline().rstrip().split()))
    m: int = int(sys.stdin.readline().rstrip())
    b: list[int] = list(map(int, sys.stdin.readline().rstrip().split()))
    for left in range(n - m + 1):
        match: bool = True
        diff: int = a[left] - b[0]
        if all(a[left + right] - b[right] == diff for right in range(1, m)):
            sys.stdout.write(f'{left + 1} ')
    sys.stdout.write('\n')


if __name__ == '__main__':
    solution()

File: Yandex/Practicum/Algorightms_and_Data_Structures/8_sprint/G._Поиск_со_сдвигом/solution.cpp
#include <iostream>
#include <vector>


void solution() {
    /*
    Time Complexity O(NM)
    Memory Complexity O(1)
    */
    int n;
    std::scanf("%d", &n);
    std::vector<int> a (n, 0);
    for (int i = 0; i < n; ++i) std::scanf("%d", &a[i]);
    int m;
    std::scanf("%d", &m);
    std::vector<int> b (m, 0);
    for (int i = 0; i < m; ++i) std::scanf("%d", &b[i]);
    for (int left = 0; left < n - m + 1; ++left) {
        bool match = true;
        int diff = a[left] - b[0];
        for (int right = 1; right < m; ++right) {
            if (a[left + right] - b[right] != diff) {
                match = false;
                break;
            }
        }
        if (match) std::printf("%d ", left + 1);
    }
    std::printf("\n");
}


int main() {
    solution();
}